lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -95,59 +96,84 @@ impl IronRDPClient {
 /// Client has an unusual lifecycle:
 /// - The function connect_rdp calls Client::new(), which creates it on the heap (Box::new), grabs a raw pointer(Box::into_raw),
 ///   and returns in to Go.
-/// - Most other exported rdp functions (pub unsafe extern "C") take the raw pointer and convert it (Box::leak(Box::from_raw(ptr)))
-///   to a reference (&'static mut Client), which can then be used without dropping the client.
+/// - Most other exported rdp functions (pub unsafe extern "C") take the raw pointer and convert it Client::from_raw
+///   to a reference (&Client), which can then be used without dropping the client.
 /// - The function free_rdp takes the raw pointer and drops it.
 ///
+/// The Client is forced to be Sync. See "Go/Rust Interface" in ../README.md for more details.
+///
 /// The Client makes use of asynchronous rust via the tokio runtime. A single runtime is created in connect_rdp and held on to by the
 /// Client. The exported rdp functions which need to call async rust functions start with `client.tokio_rt.handle().clone().block_on( ... )`,
 /// which creates a new task on the tokio runtime and blocks the current thread until it completes. Since these functions are called from
 /// Go, the "current thread" can be thought of as whichever goroutine the exported rdp function is called from. Because the client might
 /// be being used by multiple goroutines concurrently, it is up to the programmer to consider any synchronization mechanisms that might
 /// need to be implemented as features are added to the Client going forward.
 pub struct Client {
-    iron_rdp_client: IronRDPClient,
+    iron_rdp_client: Mutex<IronRDPClient>,
     tokio_rt: tokio::runtime::Runtime,
     go_ref: usize,
 }
 
+/// Forces the compiler to check that the Client struct is Send + Sync.
+/// See the README for more details.
+const _: () = {
+    const fn assert_send_sync<T: Send + Sync>() {}
+    assert_send_sync::<Client>();
+};
+
 impl Client {
     fn new(
         iron_rdp_client: IronRDPClient,
         go_ref: usize,
         tokio_rt: tokio::runtime::Runtime,
-    ) -> *mut Self {
+    ) -> *const Self {
         Box::into_raw(Box::new(Self {
-            iron_rdp_client,
+            iron_rdp_client: Mutex::new(iron_rdp_client),
             tokio_rt,
             go_ref,
         }))
     }
 
-    unsafe fn from_raw(ptr: *mut Self) -> Result<&'static mut Client, CGOErrCode> {
-        match ptr.as_ref() {
+    fn null() -> *const Self {
```

Comment:
This is used in a single place, can we inline that?

Reply:
We could, but I prefer it here per it being an expected state of the `* Client` object that ultimately gets passed around. Specifically, it's something we check for in `drop` like

```rust
    unsafe fn drop(ptr: *mut Self) {
        if !ptr.is_null() {
            drop(Box::from_raw(ptr))
        }
    }
```

I recently learned that there's a way we can get rid of all of this (h/t to our friend Benoit at Devolutions):

this signature:
pub unsafe extern "C" fn call_into_rust_from_go(client_ptr: *const Client) -> CGOErrCode {
  // snip //
}
Is equivalent to this signature (as in "same ABI"):
pub unsafe extern "C" fn call_into_rust_from_go(client: Option<&Client>) -> CGOErrCode {
  // snip //
}
client will be None when a null pointer is passed by the caller, and Some(…) for a non-null pointer. This is nice because you don’t need to call .as_ref() or in some case to use transmute. Also, you can’t forget the null check.
https://doc.rust-lang.org/nomicon/ffi.html#the-nullable-pointer-optimization
https://rust-lang.github.io/unsafe-code-guidelines/layout/function-pointers.html#representation
https://rust-lang.github.io/unsafe-code-guidelines/glossary.html#niche
Also, Box<T> , Option<Box<T> and *T also have the exact same ABI, guaranteed since Rust 1.41: https://blog.rust-lang.org/2020/01/30/Rust-1.41.0.html#more-guarantees-when-using-boxt-in-ffi
So it’s easy to implement a function to free some type by simply taking the ownership back as a Box like so:
pub unsafe extern "C" fn free_me(me: Option<Box<Me>>) {
  // freed when dropped
}

however I've left this as a TODO to myself because we need to start making more substantive progress on porting smartcard/clipboard/directory sharing. That said, it could be a good task for you to get started on this part of the codebase when you've finished what you're currently working on.